tools/resources/ignore.py

```diff
@@ -31,11 +31,46 @@ class MbedIgnoreSet(object):
 
     def __init__(self):
         self._ignore_patterns = []
-        self._ignore_regex = re.compile("$^")
+        self._ignore_regexes = []
+        self._unignore_patterns = []
+        self._unignore_regexes = []
 
     def is_ignored(self, file_path):
         """Check if file path is ignored by any .mbedignore thus far"""
-        return self._ignore_regex.match(normcase(file_path))
+
+        # find longest ignore and unignore pattern that matches the path
+        ignore_match_pattern = None
+        unignore_match_pattern = None
+
+        filepath_normcase = normcase(file_path)
+
+        for regex_index in range(0, len(self._ignore_regexes)):
+            this_regex_match = self._ignore_regexes[regex_index].match(filepath_normcase)
+            if this_regex_match:
+                if ignore_match_pattern is None:
+                    # no previous match
+                    ignore_match_pattern = self._ignore_patterns[regex_index]
+                elif len(self._ignore_patterns[regex_index]) > len(ignore_match_pattern):
+                    # found a longer match
```

Comment:
I'd prefer to see a comment explaining why a longer match is important, rather than just repeating what the code says.